src/App.tsx

```diff
@@ -1,26 +1,216 @@
+    }
+  }, [isAdding, todos.length, selectedTodo]);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const handleAddTodo = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+    const form = event.currentTarget;
+    const input = form.elements.namedItem('NewTodoField') as HTMLInputElement;
+    const title = input.value.trim();
+
+    if (!title) {
+      setError(ERROR_MESSAGES.EMPTY_TITLE);
+
+      return;
+    }
+
+    const tempId = Date.now();
+    const newTemp: Todo = {
+      id: tempId,
+      userId: USER_ID,
+      title,
+      completed: false,
+    };
+
+    setTempTodo(newTemp);
+    setProcessingIds(prev => [...prev, tempId]);
+    setIsAdding(true);
+    setError(null);
+
+    addTodo(title)
+      .then(newTodo => {
+        setTodos(current => [...current, newTodo]);
+        setTempTodo(null);
+        form.reset();
+      })
+      .catch(() => {
+        setTempTodo(null);
+        setError(ERROR_MESSAGES.ADD_TODO);
+      })
+      .finally(() => {
+        setProcessingIds(prev => prev.filter(id => id !== tempId));
+        setIsAdding(false);
+      });
+  };
+
+  const handleDeleteTodo = (todoId: number) => {
+    setError(null);
+    setProcessingIds(prev => [...prev, todoId]);
+
+    deleteTodo(todoId)
+      .then(() => {
+        setTodos(currentTodos => currentTodos.filter(t => t.id !== todoId));
+        if (selectedTodo?.id === todoId) {
+          setSelectedTodo(null);
+        }
+      })
+      .catch(() => setError(ERROR_MESSAGES.DELETE_TODO))
+      .finally(() =>
+        setProcessingIds(prev => prev.filter(id => id !== todoId)),
+      );
+  };
+
+  const handleUpdateTodo = (todoId: number, data: Partial<Todo>) => {
+    setError(null);
+    setProcessingIds(prev => [...prev, todoId]);
+
+    return updateTodo(todoId, data)
+      .then(updatedTodo => {
+        setTodos(currentTodos =>
+          currentTodos.map(t => (t.id === todoId ? updatedTodo : t)),
+        );
+        setSelectedTodo(null);
+      })
+      .catch(() => {
+        setError(ERROR_MESSAGES.UPDATE_TODO);
+      })
+      .finally(() =>
+        setProcessingIds(prev => prev.filter(id => id !== todoId)),
+      );
+  };
+
+  const handleEditTodo = (todo: Todo) => {
+    const trimmedTitle = todo.title.trim();
+
+    if (!trimmedTitle) {
+      handleDeleteTodo(todo.id);
+
+      return;
+    }
+
+    const originalTodo = todos.find(t => t.id === todo.id);
+
+    if (originalTodo && originalTodo.title !== trimmedTitle) {
+      handleUpdateTodo(todo.id, { title: trimmedTitle });
+    } else {
+      setSelectedTodo(null);
+    }
+  };
+
+  const handleToggleAll = () => {
+    setError(null);
+
+    const areAllCompleted = todos.every(todo => todo.completed);
+    const targetStatus = !areAllCompleted;
+
+    const todosToUpdate = todos.filter(todo => todo.completed !== targetStatus);
+
+    todosToUpdate.forEach(todo => {
+      handleUpdateTodo(todo.id, { completed: targetStatus });
+    });
```

Comment:
```suggestion
      todos.filter(todo => todo.completed !== targetStatus);
      .forEach(todo => {
        handleUpdateTodo(todo.id, { completed: targetStatus });
    });
```
formatting + chaining